### DISCUSSION TITLE
HTTP/2 client inside HTTP/1.1 proxy connection

### DISCUSSION THREAD

**Discussion by levshkatov** [nodejs/undici#3613] [Q&A] [answered]
I want to open a connection to proxy server and then establish H2 connection inside first one to target server. I know, that `ProxyAgent` has `allowH2` option, but it doesn't work for me, I suppose because proxy server doesn't support H2.

```js
import { Client, ProxyAgent, request } from 'undici';

const test = async () => {
  try {
    const proxy = new ProxyAgent({
      uri: `http://host:port`,
      token: `Basic ${Buffer.from(`user:pass`).toString('base64')}`,
      // allowH2: true // not working because proxy server seems not handle H2
    });

    const h2Client = new Client('https://api.ipify.org', {
      allowH2: true,
      // connect: {
      //   rejectUnauthorized: false,
      //   ...(await proxy.connect({ origin: 'http://host:port', path: '/' })), // not working
      // },
    });

    const { body, statusCode } = await request(
      `https://api.ipify.org?format=json`,
      {
        dispatcher: h2Client,
      }
    );

    console.log(statusCode);
    const data = await body.json();
    console.log(data);
  } catch (e) {
    console.log(e);
  }
};

await test();
```

I also tried to implement following [answer](https://stackoverflow.com/a/62893741) using undici but didn't get the result.
Is it possible?

**Comment by metcoder95** [comment]
Do you have the error that is thrown when attempting to connect?

**Reply by levshkatov** [reply]
I tried proxy.connect with origin as a targetHost (`https://api.ipify.org`). I didn't have errors, simply proxy was not used, website returned my local ip address. 
I wrote function that uses native nodejs api and it works properly - it opens http/1.1 tunnel to proxy and makes http/2 requests inside it
```js
const testNativeH2 = async () => {
  const req = http.request({
    method: 'CONNECT',
    host: proxyHost,
    port: proxyPort,
    path: 'api.ipify.org:443',
    headers: {
      'Proxy-Authorization': 'Basic ' + Buffer.from(`user:pass`).toString('base64'),
    },
  });
  req.end();

  req.on('connect', (res, socket) => {
    if (res.statusCode !== 200) {
      console.error(`Proxy connection failed: ${res.statusCode} ${res.statusMessage}`);
      return;
    }

    const client = http2.connect('https://api.ipify.org', {
      createConnection: () => {
        return tls.connect({
          host: 'api.ipify.org',
          servername: 'api.ipify.org',
          socket: socket,
          ALPNProtocols: ['h2'],
        });
      },
    });

    client.on('error', (err) => console.error('HTTP/2 client error:', err));

    const h2Request = client.request({
      ':path': '/?format=json',
    });

    h2Request.on('response', (headers) => {
      console.log('Status:', headers[':status']);
      let data = '';
      h2Request.on('data', (chunk) => {
        data += chunk;
      });
      h2Request.on('end', () => {
        console.log('Response:', JSON.parse(data));
        client.close();
      });
    });

    h2Request.on('error', (err) => console.error('HTTP/2 request error:', err));
    h2Request.end();
  });

  req.on('error', (err) => {
    console.error('Proxy connection error:', err);
  });
};
```

The problem is to write the same using undici abstractions.

**Reply by levshkatov** [reply]
I tried following variant, but it also returned me my ip address instead of proxy one.
```js
const testUndiciH2 = async () => {
  try {
    const proxyAgent = new ProxyAgent({
      uri: 'proxy',
      token: `Basic ${Buffer.from('user:pass').toString('base64')}`,
    });

    const { socket } = await proxyAgent.connect({
      origin: 'https://api.ipify.org:443',
      path: '/',
    });

    const h2Client = new Client('https://api.ipify.org', {
      connect: {
        rejectUnauthorized: false,
        createConnection: () => {
          servername: 'api.ipify.org',
          socket
        },
      },
      allowH2: true,
    });

    const { statusCode, body } = await h2Client.request({
      path: '/?format=json',
      method: 'GET',
    });

    console.log('Status:', statusCode);
    const data = await body.json();
    console.log('Response:', data);
    await h2Client.close();
  } catch (error) {
    console.error('Error:', error);
  }
};
```

**Reply by levshkatov** [reply]
It seems that undici doesn't use my custom socket and creates default one
https://github.com/nodejs/undici/blob/3d0ce67ff75ec061497d6092e5820b6721f6de66/lib/core/connect.js#L89

**Reply by metcoder95** [reply]
That's slightly more complex, another way (and maybe better) it is to do something similar as we do for `ProxyAgent`, where an `Agent` is used to manage that upgrade.

See: https://github.com/nodejs/undici/blob/0ad6007ae74ce2d392c76a90b07e6cda725d3bad/lib/dispatcher/proxy-agent.js#L63

Which overall to override the whole `connect` property.

**Reply by levshkatov** [reply]
Thank you for your answer, good point to start but it looks a bit difficult to me because of lack of docs about connect and socket upgrade. I tried following code, but got an error
```js
const proxyAgent = new ProxyAgent({
  uri: 'proxy',
  token: `Basic ${Buffer.from('user:pass').toString('base64')}`,
});

const connector = buildConnector({ allowH2: true });

const h2 = new Agent({
  allowH2: true,
  connect: async (opts, callback) => {
    try {
      const { socket } = await proxyAgent.connect({
        origin: 'https://api.ipify.org',
        path: '/',
      });
      connector({ ...opts, servername: 'api.ipify.org', httpSocket: socket }, callback);
    } catch (e) {
      if (err.code === 'ERR_TLS_CERT_ALTNAME_INVALID') {
        // Throw a custom error to avoid loop in client.js#connect
        callback(new SecureProxyConnectionError(err));
      } else {
        callback(err);
      }
    }
  },
});

const { statusCode, headers, body } = await request('https://api.ipify.org/?format=json', {
  dispatcher: h2,
});
```

```
Error: Error: 40284C41517F0000:error:0A00010B:SSL routines:ssl3_get_record:wrong version number:../deps/openssl/openssl/ssl/record/ssl3_record.c:354:

    at initRead (node:_tls_wrap:502:25)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  library: 'SSL routines',
  reason: 'wrong version number',
  code: 'ERR_SSL_WRONG_VERSION_NUMBER'
}
```

I'm a bit confused in all these path, origin, servername. Maybe I provide wrong info for any of it. Could you help?

**Reply by levshkatov** [reply]
My main idea is to build one agent on another (h2 based on proxyAgent), but it is impossible to do by simply passing it like a dispatcher. I understand that it's possible to make via connect function but I'm too new into undici for understanding all machinery there.

**Reply by metcoder95** [reply, marked as answer]
Yeah, see what you mean. Here what you want to do is to first connect to the Proxy and use that socket to do the requests in the second agent/dispatcher.

What I was suggesting was not to use the Agent and have an internal `connect` that goes to a Proxy, but actually the other way around.

On the `ProxyAgent` you pass a custom connect that connects to the proxy with H1 (using an Agent) and returns that socket to the `connect` callback so it can be reused across request. Similar to what ProxyAgent does to connect to the Proxy.

> The error you are receiving seems to be related with you connecting with `https` but the proxy supporting only `http`.

**Reply by levshkatov** [reply]
Thanks a lot for explanation. It helped me rewrite `ProxyAgent` as a custom one, it worked properly and then I realized that if I simply pass `requestTls: {allowH2: true}` to `ProxyAgent` options it will do the same. 
First time I tried to pass `allowH2` without `requestTls`, on the first level of options and it didn't work for sure, because my proxy doesn't support H2. Reading source code could be helpful sometimes, haha. Thanks again!
